src/librustc_middle/ty/walk.rs

```diff
@@ -3,7 +3,54 @@
 
 use crate::ty;
 use crate::ty::subst::{GenericArg, GenericArgKind};
+use arrayvec::ArrayVec;
+use rustc_data_structures::fx::FxHashSet;
 use smallvec::{self, SmallVec};
+use std::hash::Hash;
+
+/// Small-storage-optimized implementation of a set
+/// made specifically for walking type tree.
+///
+/// Stores elements in a small array up to a certain length
+/// and switches to `HashSet` when that lenght is exceeded.
+pub enum MiniSet<T: Eq + Hash + Copy> {
+    Array(ArrayVec<[T; 8]>),
+    Set(FxHashSet<T>),
+}
+
+impl<T: Eq + Hash + Copy> MiniSet<T> {
+    /// Creates an empty `MiniSet`.
+    pub fn new() -> Self {
+        MiniSet::Array(ArrayVec::new())
+    }
+
+    /// Adds a value to the set.
+    ///
+    /// If the set did not have this value present, true is returned.
+    ///
+    /// If the set did have this value present, false is returned.
+    pub fn insert(&mut self, elem: T) -> bool {
+        match self {
+            MiniSet::Array(array) => {
+                if array.iter().any(|e| *e == elem) {
+                    false
+                } else {
+                    if !array.is_full() {
+                        unsafe {
+                            array.push_unchecked(elem);
+                        }
+                    } else {
+                        let mut set: FxHashSet<T> = array.iter().copied().collect();
+                        set.insert(elem);
+                        *self = MiniSet::Set(set);
+                    }
```

Comment:
Is it really worth using an unsafe operation here?

How about something like:

```suggestion
                    if let Err(cap_err) = array.try_push(elem) {
                        let elem = cap_err.element();
                        let mut set: FxHashSet<T> = array.iter().copied().collect();
                        set.insert(elem);
                        *self = MiniSet::Set(set);
                    }
```

Reply:
Agreed, I'd prefer something like this as well.

Reply:
Great change. Didn't know about try_push. Will update.